workspaces/mcp-chat/plugins/mcp-chat-backend/src/utils.ts

```diff
@@ -0,0 +1,405 @@
+/*
+ * Copyright 2025 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { spawn } from 'child_process';
+import * as path from 'path';
+import { promises as fs } from 'fs';
+import { Client } from '@modelcontextprotocol/sdk/client/index.js';
+import { ServerConfig, VALID_ROLES } from './types';
+import { RootConfigService } from '@backstage/backend-plugin-api';
+
+export function loadServerConfigs(config: RootConfigService): ServerConfig[] {
+  const mcpServers = config.getOptionalConfigArray('mcpChat.mcpServers') || [];
+
+  return mcpServers?.map(serverConfig => {
+    const headers: Record<string, string> | undefined = serverConfig
+      .getOptionalConfig('headers')
+      ?.get() as Record<string, string> | undefined;
+
+    const env: Record<string, string> | undefined = serverConfig
+      .getOptionalConfig('env')
+      ?.get() as Record<string, string> | undefined;
+
+    const type =
+      serverConfig.getOptionalString('type') || serverConfig.has('url')
+        ? 'streamable-http'
+        : 'stdio';
+
+    return {
+      id: serverConfig.getString('id'),
+      name: serverConfig.getString('name'),
+      scriptPath: serverConfig.getOptionalString('scriptPath'),
+      npxCommand: serverConfig.getOptionalString('npxCommand'),
+      args: serverConfig.getOptionalStringArray('args'),
+      env,
+      url: serverConfig.getOptionalString('url'),
+      headers,
+      type,
+    };
+  });
+}
+
+/**
+ * Helper function to find npx executable path
+ * Searches common installation locations and validates the executable works
+ * @returns Promise<string> Path to the npx executable
+ * @throws Error if npx cannot be found or is not functional
+ */
+export async function findNpxPath(): Promise<string> {
+  // Get the directory where node is installed
+  const nodeDir = path.dirname(process.execPath);
+
+  const possiblePaths = [
+    'npx', // Try system PATH first
+    path.join(nodeDir, 'npx'), // Same dir as node (Unix)
+    path.join(nodeDir, 'npx.cmd'), // Windows
+    '/usr/local/bin/npx', // Common installation path
+    '/opt/homebrew/bin/npx', // Homebrew on Apple Silicon
+  ];
```

Comment:
My opinion: People should just configure their system correct. I would avoid this kind of "magic", but its up to you. :)